### PR TITLE
Make Codensity kind and levity polymorphic

### PR DESCRIPTION
**Pull request by treeowl**
With `TypeInType`, we can get

```haskell
Codensity :: (k -> TYPE rep) -> Type -> Type
```

I don't know if the polymorphism in `k` is useful; it was
readily available, so I took it. I actually have an application in
mind for polymorphism in `rep`. The `primitive` package offers

```haskell
indexArrayM :: Monad m => Array a -> Int -> m a
```

Now what if I want to recover the underlying functionality of
`indexArray#` from that? One option would be to use the
`data Box a = Box a` "lifted identity" monad. Another would be
to use a custom-written `Cont` monad (a clone of the one from
`transformers`). But I realized that `Codensity` *almost* works off
the shelf, if we just add a bit of extra polymorphism. With this patch,
I can write

```haskell
iarr# :: Array a -> Int -> (# a #)
iarr# ar i = runCodensity (indexArrayM ar i) (\a -> (# a #))
```

and get exactly the Core I want

```
iarr# :: forall a. Array a -> Int -> (# a #)
[GblId,
 Arity=2,
 Caf=NoCafRefs,
 Str=<S(S),1*U(U)><S(S),1*U(U)>,
 Unf=Unf{Src=InlineStable, TopLvl=True, Value=True, ConLike=True,
         WorkFree=True, Expandable=True,
         Guidance=ALWAYS_IF(arity=2,unsat_ok=True,boring_ok=False)
         Tmpl= \ (@ a_aetV)
                 (ar_aetE [Occ=Once!] :: Array a_aetV)
                 (i_aetF [Occ=Once!] :: Int) ->
                 case i_aetF of { GHC.Types.I# i#_aeun [Occ=Once] ->
                 case ar_aetE of { Array ds2_aeur [Occ=Once] ->
                 GHC.Prim.indexArray# @ a_aetV ds2_aeur i#_aeun
                 }
                 }}]
iarr#
  = \ (@ a_aetV) (ar_aetE :: Array a_aetV) (i_aetF :: Int) ->
      case i_aetF of { GHC.Types.I# i#_aeun ->
      case ar_aetE of { Array ds2_aeur ->
      GHC.Prim.indexArray# @ a_aetV ds2_aeur i#_aeun
      }
      }
```

### CODE REVIEW

**Comment by treeowl** [comment]
By the way: it is also quite possible to make `ContT` work with this sort of thing. Unfortunately, that means convincing Ross Paterson to do it.

**Comment by ekmett** [comment]
The levity polymorphism seems to work out to be free.

On the other hand, the `PolyKinds` can wind up with many potential instances becoming flexible

Consider the following fairly artificial example, 

```haskell
instance (Applicative f, Monoid m) => Monoid (Codensity f m)
```

It wasn't flexible before `PolyKinds` were added, but afterwards, (as shown by haddock)

```haskell
instance (Applicative f, Monoid m) => Monoid (Codensity * f m)
```

is a flexible instance. So until something monomorphizes the kind of the argument, the instance isn't selectable, leading to more situations where instance selection gets stuck.

At present, IIRC, we already have `PolyKinds` turned on, so we've already paid this latter cost.

**Comment by treeowl** [comment]
Good point. Perhaps we should unflex the instances by fixing the kind in the constraint where appropriate?

**Comment by ekmett** [comment]
The reason I accepted the original proposal to adopt `PolyKinds` for this data type was that as I recall, none of the existing instances happened to run afoul of this issue, but its a thing I consider whenever I get an enthusiastic email asking me to PolyKind all the things.

File: src/Control/Monad/Codensity.hs [hunk 1]
```diff
@@ -47,9 +51,12 @@ import Data.Functor.Apply
 import Data.Functor.Kan.Ran
 import Data.Functor.Plus
 import Data.Functor.Rep
-#if __GLASGOW_HASKELL__ >= 708
+#if (__GLASGOW_HASKELL__ >= 708) && (__GLASGOW_HASKELL__ < 800)
 import Data.Typeable
 #endif
+#if __GLASGOW_HASKELL__ >= 802
+import GHC.Exts (TYPE)
```
Comment (RyanGlScott) [inline src/Control/Monad/Codensity.hs]: Levity polymorphism has been around since GHC 8.0—is there a reason you only enabled it on 8.2 or later?

**Comment by treeowl** [comment]
Maybe I got the version wrong...

**Comment by treeowl** [comment]
@RyanGlScott, FYI, in a new PR, I tried 8.0 and the compiler panicked. `TypeInType` was pretty experimental then.

**Comment by RyanGlScott** [comment]
Ah! Thanks for confirming.

**Comment by Icelandjack** [comment]
@ekmett Is this at all better than the flexible version?

```haskell
instance (f ~~ f', Applicative f', Monoid m) => Monoid (Codensity (f::k -> Type) m)
```

**Comment by treeowl** [comment]
I submitted a PR suggesting this approach. I think it works for everything
except MonadTrans.

**Comment by Icelandjack** [comment]
Great! #53

**Pull request merged**